engines/twp/scenegraph.cpp

```diff
@@ -0,0 +1,844 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "math/matrix3.h"
+#include "common/algorithm.h"
+#include "common/config-manager.h"
+#include "twp/twp.h"
+#include "twp/scenegraph.h"
+#include "twp/lighting.h"
+
+namespace Twp {
+
+#define DEFAULT_FPS 10.f
+
+#define NUMOBJECTSBYROW 4
+#define MARGIN 8.f
+#define MARGINBOTTOM 10.f
+#define BACKOFFSET 7.f
+#define ARROWWIDTH 56.f
+#define ARROWHEIGHT 86.f
+#define BACKWIDTH 137.f
+#define BACKHEIGHT 75.f
+
+static float _getFps(float fps, float animFps) {
+	if (fps != 0.f)
+		return fps;
+	return (animFps < 1e-3) ? DEFAULT_FPS : animFps;
+}
+
+Node::Node(const Common::String &name, Math::Vector2d scale, Color color)
+	: _name(name),
+	  _color(color),
+	  _computedColor(color),
+	  _scale(scale) {
+}
+
+Node::~Node() {
+	remove();
```

Comment:
Correct, I feel stupid I warned the parent to remove the node but not the children.
That's crazy because I've done it so much time in my daily job.

Reply:
Tested your last change and now I'm getting regular sigtraps in the tutorial slideshow. Something still seems to be wrong.

Reply:
Yeah sorry about that, I will fix it, I pushed it too fast.